File: C++/BIT.cpp
#include<iostream>
#include<vector>
using namespace std;


using ll=long long;

template<class T> struct BIT{
using VT=vector<T>;
T n;
VT l;
BIT(const T& num):n(num),l(n+1,0){}
BIT(const T& num,const T& init_l):n(num),l(n+1,init_l){}
T lsb(T i){return i&(-i);}
T add(T i, T j){//add j to l_i
T e=i;
while(e<n+1){
    l.at(e)+=j;
    e+=lsb(e);
}
return j;
}
T update(T i,T j){//update j to l_i
return add(i,j-l.at(i))+j;
}
ll sum(T le,T ri){
return sum(ri)-sum(le-1);
}
ll sum(T le){
    ll ans=0;
    while(le>0){
        ans+=l.at(le);
        le-=lsb(le);
    }
    return ans;
}
};
int main(){
    
}

File: python/BIT.py
import random


class BIT:
    def __init__(self, n: int, v: int = 0):
        self.n = n
        self.l = [v for i in range(n+1)]

    def __len__(self) -> int:
        return self.n

    @classmethod
    def lsb(cls, i) -> int:
        return i & (-i)

    def update(self, i: int, j: int):
        return self.add(i, j-self.l[i])

    def add(self, i: int, j: int):
        k = j
        e = i
        while e < self.n+1:  # 1-indexed
            self.l[e] += k
            e += BIT.lsb(e)

    def sum(self, right: int, left: int = None) -> int:
        if left:
            return self.sum(left)-self.sum(right-1)
        self.raiseIndexError(right)
        ans = 0
        while right:
            ans += self.l[right]
            right -= BIT.lsb(right)
        return ans

    def __iter__(self):
        for i in self.l:
            yield i

    def __str__(self):
        return str(list(self))

    def __setitem__(self, key, value):
        self.raiseIndexError(key)
        self.update(key, value)

    def __getitem__(self, key):
        self.raiseIndexError(key)
        return self.l[key]

    def isIndexError(self, key):
        return key > self.n+1

    def raiseIndexError(self, key):
        if self.isIndexError(key):
            raise IndexError('list index out of range')


def checkBIT():
    b = BIT(10)
    print(b)
    for i in range(int(input())):
        u, t, f = map(int, input().split())
        if u == 1:
            b[t] = f
        else:
            print(b.sum(t, f))
    print(b, b[2])
    print(b.isIndexError(100))


def main():
    n = int(input())
    b = BIT(n)
    l = list(map(int, input().split()))
    ans = -n
    for i in l:
        b.add(i, 1)
        ans += b.sum(i, n)
    print(ans)


if __name__ == '__main__':
    main()
